Support Amazon S3 storage

skarllot opened the pull request:
Implement file-system interfaces using  Amazon S3 storage.

emgarten commented:
S3 support!! @skarllot you rock!! 🥇 🍰 💯 

At first glance this looks great, glad to see all the tests and end to end tests here! I'm going to test this out myself and review it a bit closer later tonight or tomorrow.

emgarten commented on src/SleetLib/FileSystem/AmazonS3FileSystemLock.cs:
```diff
@@ -0,0 +1,112 @@
+                    if (!await FileExistsAsync(client, bucketName, LockFile, token).ConfigureAwait(false))
+                    {
+                        string fileContent = DateTime.UtcNow.ToString("O", DateTimeFormatInfo.InvariantInfo);
+                        await CreateFileAsync(client, bucketName, LockFile, fileContent, token).ConfigureAwait(false);
```
Does this fail if the file already exists? It would be a short window of time between the exists check and when this is written, but I'm wondering if it needs to do more to avoid race conditions?

skarllot replied:
No, it does not fail. I didn't found a way to avoid race condition using purely the S3.
How important the file system locking is?
I could achieve that using the Amazon messaging service.

emgarten replied:
It doesn't seem worth adding a new service beyond s3 just for locking, this should get pretty close.

emgarten commented on src/SleetLib/FileSystem/AmazonS3FileSystemLock.cs:
```diff
@@ -0,0 +1,112 @@
+using System.Threading.Tasks;
+using Amazon.S3;
+using NuGet.Common;
+using static Sleet.AmazonS3FileSystemAbstraction;
```
Did not realize this was possible 😄

skarllot replied:
Yes, it is since version 6.0 of C#. Several cool features was introduced to C# on latest versions.

emgarten commented on src/SleetLib/FileSystem/AmazonS3FileSystem.cs:
```diff
@@ -0,0 +1,97 @@
+            return new AmazonS3File(this, rootUri, uri, LocalCache.GetNewTempPath(), client, bucketName, key);
+        }
+
+        private string GetPathRelativeToBucket(Uri uri)
```
Is it possible to support folders within a bucket? In the azure file system for sleet it is possible to append an extra part of the path, this allows multiple feeds within the same bucket.

skarllot replied:
Yes, you use forward slashes on objects names.

emgarten commented:
Overall this looks great! I was able to make a feed on s3 and use it from Visual Studio once I figured out how to create a user and set up the access policies 🎉

skarllot commented:
> Sleet.exe (ILMerged) doesn't seem to contain the AWS assemblies and fails when using an s3 feed. Haven't had a chance to investigate that yet.

Fixed. I've added Amazon assemblies to ILMerge building.

> Is the user expected to set the bucket to public read only through the AWS console?

I'm not sure what you mean.

skarllot commented:
The ILMerge is not add the 'System.Reflection.Metadata.dll' assembly too.

emgarten commented:
> The ILMerge is not adding the 'System.Reflection.Metadata.dll' assembly too.

I'll take a look, it has been a lot of trial and error getting the merge list right for some reason.

skarllot commented:
> I'll take a look, it has been a lot of trial and error getting the merge list right for some reason.
Is it harmful to add every assemblies (*.dll)?

emgarten commented:
> Is it harmful to add all assemblies (*.dll)?

It would contain a lot of *system.** dlls that are part of the .NET Framework, it would end up making the file size bigger than it needs to be.

emgarten commented:
Docs look good, thanks for adding that!

emgarten commented:
> Is it harmful to add all assemblies (*.dll)?

I think you are correct, it just needs everything in the bin folder. This may have changed in how dotnet CLI handles this, in the past it was copying every System.*.dll into the folder. Now it looks like just the needed ones are present. I've updated the script, let me know if still doesn't work for you.
